### DISCUSSION TITLE
Discussion: FILE_UPLOAD=local 環境下での Attachment 保存場所の予期しない変更(bug 含む)への対処について

### DISCUSSION THREAD

**Discussion by hanapage** [growilabs/growi#6086] [Help / Questions] [answered]
## Informations
**Is your feature request related to a problem? Please describe.**
- ローカルでビルド/添付ファイルを保存しているユーザ向けのリクエストです。
- GROWIのファイルアップロード先をLocalに指定している時、仕様変更でpublicの場所が変わってしまうと別の場所に新規にpublicディレクトリが作成され、過去に作られたpublicディレクトリは参照されなくなってしまう。そのため、添付ファイルへのリンクが切れてしまい画像などが表示されなくなる。
- この問題自体は過去のpublicの内容を新しいpublicにコピーすることで解決できるが、そもそもpublicが移動したことに気が付けなければ問題を自己解決することができない。(実際Twitterで数人のユーザがこの問題にぶつかっていたのを確認しています)
- おそらく過去2回はattachmentの場所が変更されていると思います。今後も変更があるのであれば教えてほしいです。

**Describe the solution you'd like**
- Localの添付ファイルの保存先が仕様変更された時にアップグレードガイド又はリリースノートで対応方法を告知してほしい。

**Describe alternatives you've considered**
- 過去のpublicにある添付ファイルも参照するようにしてほしい。
- もしくはpublicの場所を環境変数などで指定できるようにしてほしい。

いつも便利に使わせていただいてます。よろしくお願いします。

**Comment by yuki-takei** [comment]
@hanapage ご意見ありがとうございます。

FILE_UPLOAD=local の設定をしている環境に対して、テストやケアが十分ではないのは事実で、我々としても申し訳なく思っています。

> おそらく過去2回はattachmentの場所が変更されていると思います。

に関しては、開発チームではバージョンアップに伴うデグレと捉えていますが、そもそもデグレが起こることを予見できていませんでした。(そのため、アップグレードガイドに載せようという話にそもそもなっていない)

また、今現在もどういった症状だったのかというのを整理する(正確に把握する)のが難しく、掲載に至っていません。


## 今後の対応 1

> Localの添付ファイルの保存先が仕様変更された時にアップグレードガイド又はリリースノートで対応方法を告知してほしい。

関知できる場合はもちろんそうします。また、今からでも以下の情報が揃えば docs を更新します。

1. from version と to version
2. 症状
3. ワークアラウンド

## 今後の対応 2

> 過去のpublicにある添付ファイルも参照するようにしてほしい。

こちらも情報が揃えば検討いたします。

> もしくはpublicの場所を環境変数などで指定できるようにしてほしい。

こちらは今のところ対応予定はありません。
PR をいただければ取り込むのは可能です。

## その他

当該環境向けに全体的に塩対応な感じになってしまうのは申し訳ないです。。

**Comment by hanapage** [comment]
レスポンスありがとうございます。
どのバージョンで変わったのか正確には把握できてませんが、私の環境でタイムスタンプを見る限りでは

- /opt/growi/attachment/public (導入時=2018/09頃)
- /opt/growi/public (2019/10頃～)
- /opt/growi/packages/app/public (2021/09頃～)

の順に変わっていると思います。

**Comment by yamatomo717** [comment, marked as answer]
横から失礼致します。

私の方でも
`/opt/growi/public` から `/opt/growi/packages/app/public` に
変わった際にトラブルがございましたので情報提供致します。

1. from version と to version

from version: 4.3.3 `/opt/growi/public`
to verstion: 4.4.0 `/opt/growi/packages/app/public`

2. 症状

> GROWIのファイルアップロード先をLocalに指定している時、仕様変更でpublicの場所が変わってしまうと別の場所に新規にpublicディレクトリが作成され、過去に作られたpublicディレクトリは参照されなくなってしまう。そのため、添付ファイルへのリンクが切れてしまい画像などが表示されなくなる。

こちらの通りです。
4.3.3 以前でローカルに保存した画像( /opt/growi/public/uploads に保存した画像)が表示されなくなります。

3. ワークアラウンド

私の環境では、
`/opt/growi/public/uploads` を `/opt/growi/packages/app/public/uploads` に移動することで解決しました。

**Comment by yuki-takei** [comment]
## v6.1.0 での変更
2023年5月にリリースを目指している v6.1.0 で、保存場所が再び変更されますのでまずは頭出しです。

### Before -> After
`/opt/growi/packages/app/public` -> `/opt/growi/apps/app/public`

## 今後の対応
v6.1.0 へのアップグレードガイド([en](https://docs.growi.org/en/admin-guide/upgrading/61x.html)/[ja](https://docs.growi.org/ja/admin-guide/upgrading/61x.html))に載せる予定です
